typing/typecore.ml

```diff
@@ -1682,6 +1685,649 @@ and is_nonexpansive_opt = function
     None -> true
   | Some e -> is_nonexpansive e
 
+module Env' = Env
+module Rec_context =
+struct
+  type access =
+      Dereferenced
+    (** [Dereferenced] indicates that the value (not just the address) of a
+        variable is accessed *)
+
+    | Guarded
+    (** [Guarded] indicates that the address of a variable is used in a
+        guarded context, i.e. under a constructor.  A variable that is
+        dereferenced within a function body or lazy context is also considered
+        guarded. *)
+
+    | Unguarded
+    (** [Unguarded] indicates that the address of a variable is used in an
+        unguarded context, i.e. not under a constructor. *)
+
+  (** [guard] represents guarded contexts such as [C -] and [{l = -}] *)
+  let guard : access -> access = function
+    | Dereferenced -> Dereferenced
+    | Guarded -> Guarded
+    | Unguarded -> Guarded
+
+  (** [inspect] represents elimination contexts such as [match - with cases],
+      [e -] and [- e] *)
```

Comment:
My first reaction when reading this was that the typing derivation

```
x:guarded |- (x, x) ok
-----------------------------------------------
x:dereferenced |- match (x, x) with (z, y) -> e
```

was a bit imprecise (we could see that the access of `x` is not made Dereferenced by this match, it should take the access mode of `z, y` in `e`). But in fact I think this is just fine, given that there is no reason for users to write those beta-redexes anyway as they can easily reduce the safe ones away.

Reply:
I would still add a comment:

```
Note that `inspect` can lose precision on terms such as `match [x] with [] -> .. | _ -> ...`, as `x` is not actually eliminated, but there is no reason for users to write those terms as they can always be reduced to a simpler form. When the scrutinee of a `match` is not reducible anymore, any dereference of `x` in it may potentially be the value being scrutinized.
```

Reply:
I've added this comment.

(I'm going to respond to the review before pushing the latest update, since GitHub sometimes makes it trickier to respond to comments after a force-push.  The changes themselves should appear shortly.)